newa/__init__.py

```diff
@@ -968,3 +998,194 @@ def drop_obsoleted_issue(self, issue: Issue, obsoleted_by: Issue) -> None:
                                              transition=self.transitions["dropped"][0])
         except jira.JIRAError as e:
             raise Exception(f"Cannot close issue {issue}!") from e
+
+#
+# ReportPortal communication
+#
+
+
+@define
+class ReportPortal:
+
+    token: str
+    url: str
+    project: str
+
+    def get_launch_url(self, launch_id: str) -> str:
+        return f"{self.url}/ui/#{self.project}/launches/all/{launch_id}"
+
+    def get_request(self,
+                    path: str,
+                    params: Optional[dict[str, str]] = None,
+                    version: int = 1) -> JSON:
+        base_url = f'{self.url}/api/v{version}/{self.project}/{path.lstrip("/")}'
+        get_params = '&'.join([f'{k}={v}' for (k, v) in params.items()]) if params else ''
+        url = f'{base_url}?{get_params}'
+        headers = {"Authorization": f"bearer {self.token}", "Content-Type": "application/json"}
+        req = requests.get(url, headers=headers)
+        if req.status_code == 200:
+            return req.json()
+        return None
+
+    def post_request(self,
+                     path: str,
+                     json: JSON,
+                     version: int = 1) -> JSON:
+        url = f'{self.url}/api/v{version}/{self.project}/{path.lstrip("/")}'
+        headers = {"Authorization": f"bearer {self.token}", "Content-Type": "application/json"}
+        req = requests.post(url, headers=headers, json=json)
+        if req.status_code == 200:
+            return req.json()
+        return None
+
+    def find_launches_by_attr(self, attr: str, value: str) -> list[str]:
+        """ Searches for RP launching having the respective attribute=value set. """
+        path = '/launch'
+        params = {'filter.has.compositeAttribute': f'{attr}:{value}'}
+        data = self.get_request(path, params)
+        if not data:
+            return []
+        return [launch['id'] for launch in data['content']]
+
+    def merge_launches(self,
+                       launch_ids: list[str],
+                       launch_name: str,
+                       description: str,
+                       attributes: Optional[dict[str, str]] = None) -> str:
+        query_data: JSON = {
+            "attributes": [],
+            'description': description,
+            'name': launch_name,
+            "mergeType": "BASIC",
+            'mode': 'DEFAULT',
+            "extendSuitesDescription": 'false',
+            "launches": launch_ids,
+            }
+        if attributes:
+            for key, value in attributes.items():
+                query_data['attributes'].append({"key": key.strip(), "value": value.strip()})
+        print(f'Merging launches: {launch_ids}')
+        data = self.post_request('/launch/merge', json=query_data)
+        if data:
+            print('Launches successfully merged')
+            return str(data['id'])
+        print('Failed to merge launches')
+        return ''
```

Comment:
It's safer to use `None` than an empty string for error-like return values. It's then easier for linters to follow them around. Consider how this return value is used:

```python
            if len(to_merge) > 1:
                final_launch = rp.merge_launches(
                    to_merge, jira_launch_name_mapping[jira_id], description, {})
            else:
                final_launch = to_merge[0]
            launch_url = rp.get_launch_url(final_launch)
```

If `merge_launches()` fails to merge launches, it'd return an empty string which would then be passed to `get_launch_url()`, leading to "launch URL" `{self.url}/ui/#{self.project}/launches/all/`. I don't know whether such an URL is expected, or whether an empty string is propagated on purpose, but that's the point: it's a tricky value, one that looks like an error is being reported, then it's used like nothing happened. If `None` would be returned instead, there would have to be special handling for `final_launch` being `None` - is it an error, or do we get URL that lists all launches, and so on. An empty string can easily spread doubt through the code.

Reply:
Fixed. I have rewritten launch merging a bit.